code/modules/mod/mod_modules.dm

```diff
@@ -0,0 +1,607 @@
+/obj/item/mod/module
+	name = "MOD module"
+	icon_state = "module"
+	/// If it can be removed
+	var/removable = TRUE
+	/// If it's passive, active or usable
+	var/module_type = MODULE_PASSIVE
+	/// Is the module active
+	var/active = FALSE
+	/// How much space it takes up in the MOD
+	var/complexity = 0
+	/// Power use when idle
+	var/idle_power_cost = 0
+	/// Power use when active
+	var/active_power_cost = 0
+	/// Power use when used
+	var/use_power_cost = 0
+	/// ID used by their TGUI
+	var/tgui_id = null
+	/// Can we be configured
+	var/configurable = FALSE
+	/// Linked MODsuit
+	var/obj/item/mod/control/mod
+	/// If we're an active module, what item are we?
+	var/obj/item/device
+	/// Overlay given to the user when the module is inactive
+	var/overlay_state_inactive
+	/// Overlay given to the user when the module is active
+	var/overlay_state_active
+	/// What modules are we incompatible with?
+	var/list/incompatible_modules = list()
+	/// Cooldown after use
+	var/cooldown_time = 0
+	/// Timer for the cooldown
+	COOLDOWN_DECLARE(cooldown_timer)
+
+/obj/item/mod/module/Initialize()
+	. = ..()
+	if(module_type != MODULE_ACTIVE)
+		return
+	if(ispath(device))
+		device = new device(src)
+		ADD_TRAIT(device, TRAIT_NODROP, MOD_TRAIT)
+		RegisterSignal(device, COMSIG_PARENT_PREQDELETED, .proc/on_device_deletion)
+		RegisterSignal(src, COMSIG_ATOM_EXITED, .proc/on_exit)
+
+/obj/item/mod/module/Destroy()
+	if(mod)
+		mod.uninstall(src)
+	if(device)
+		UnregisterSignal(device, COMSIG_PARENT_PREQDELETED)
+		QDEL_NULL(device)
+	..()
+
+/obj/item/mod/module/proc/on_install()
+	return
+
+/obj/item/mod/module/proc/on_uninstall()
+	return
+
+/obj/item/mod/module/proc/on_equip()
+	return
+
+/obj/item/mod/module/proc/on_unequip()
+	return
+
+
+/obj/item/mod/module/proc/on_select()
+	if(!mod.active || mod.activating)
+		return
+	if(module_type != MODULE_USABLE)
+		if(active)
+			on_deactivation()
+		else
+			on_activation()
+	else
+		on_use(mod.wearer)
+
+/obj/item/mod/module/proc/on_activation()
+	if(!COOLDOWN_FINISHED(src, cooldown_timer))
+		return FALSE
+	if(!mod.active || mod.activating || !mod.cell?.charge)
+		return FALSE
+	active = TRUE
+	if(module_type == MODULE_ACTIVE)
+		if(mod.selected_module)
+			mod.selected_module.on_deactivation()
+		mod.selected_module = src
+		if(device)
+			mod.wearer.put_in_hands(device)
+			to_chat(mod.wearer, "<span class='notice'>You extend [device].</span>")
+			RegisterSignal(mod.wearer, COMSIG_ATOM_EXITED, .proc/on_exit)
+		else
+			to_chat(mod.wearer, "<span class='notice'>You activate [src]. You can use the middle-click button to use it.</span>")
```

Comment:
This is bad UX, laptops don't have a middle-click button for instance. Choose a different key, perhaps an action button.